bluebutton-data-pipeline-rif-extract/src/main/java/gov/hhs/cms/bluebutton/datapipeline/rif/extract/s3/DataSetMonitorWorker.java

```diff
@@ -0,0 +1,283 @@
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3Client;
+import com.amazonaws.services.s3.model.DeleteObjectsRequest;
+import com.amazonaws.services.s3.model.DeleteObjectsRequest.KeyVersion;
+import com.amazonaws.services.s3.model.GetObjectRequest;
+import com.amazonaws.services.s3.model.ListObjectsRequest;
+import com.amazonaws.services.s3.model.ObjectListing;
+import com.amazonaws.services.s3.model.S3Object;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+
+import gov.hhs.cms.bluebutton.datapipeline.rif.extract.s3.DataSetManifest.DataSetManifestEntry;
+import gov.hhs.cms.bluebutton.datapipeline.rif.model.RifFile;
+import gov.hhs.cms.bluebutton.datapipeline.rif.model.RifFilesEvent;
+
+/**
+ * <p>
+ * Acts as a worker {@link Runnable} for {@link DataSetMonitor}. It is expected that
+ * this will be run on a repeating basis, via a {@link ScheduledExecutorService}
+ * .
+ * </p>
+ * <p>
+ * When executed via {@link #run()}, the {@link DataSetMonitorWorker} will scan the
+ * specified Amazon S3 bucket. It will look for <code>manifest.xml</code>
+ * objects/files and select the oldest one available. If such a manifest is
+ * found, it will then wait for all of the objects in the data set represented
+ * by it to become available. Once they're all available, it will kick off the
+ * processing of the data set, and block until that processing has completed.
+ * </p>
+ */
+final class DataSetMonitorWorker implements Runnable {
+	private static final Pattern REGEX_MANIFEST = Pattern.compile("^(.*)\\/manifest\\.xml$");
+
+	private final String bucketName;
+	private final DataSetProcessor dataSetProcessor;
+	private final AmazonS3 s3Client = new AmazonS3Client(new DefaultAWSCredentialsProviderChain());
+
+	/**
+	 * Constructs a new {@link DataSetMonitorWorker} instance.
+	 * 
+	 * @param bucketName
+	 *            the name of the AWS S3 bucket to monitor
+	 * @param dataSetProcessor
+	 *            the {@link DataSetProcessor} responsible for processing each
+	 *            of the {@link RifFilesEvent}s that are generated by this
+	 *            {@link DataSetMonitorWorker}
+	 */
+	public DataSetMonitorWorker(String bucketName, DataSetProcessor dataSetProcessor) {
+		this.bucketName = bucketName;
+		this.dataSetProcessor = dataSetProcessor;
+	}
+
+	/**
+	 * @see java.lang.Runnable#run()
+	 */
+	@Override
+	public void run() {
+		/*
+		 * Request a list of all objects in the configured bucket and directory.
+		 * (In the results, we'll be looking for the oldest manifest file, if
+		 * any.)
+		 */
+		ListObjectsRequest bucketListRequest = new ListObjectsRequest();
+		bucketListRequest.setBucketName(bucketName);
+
+		/*
+		 * S3 will return results in separate pages. Loop through all of the
+		 * pages, looking for the oldest manifest file that is available.
+		 */
+		String manifestToProcessKey = null;
+		ObjectListing objectListing = s3Client.listObjects(bucketListRequest);
```

Comment:
S3's data architecture is a bit unusual:

* Buckets are equivalent to file shares. In our case, there's just one bucket/share that all data sets will get pushed to.
* Within buckets, you can have objects, which have (text) keys.
* By convention, keys use '`/`' characters to denote directories. If I push a file `foo.txt` to S3 and assign it a key of '`bar/foo.txt`', then S3's graphical interface will display it as if that `foo.txt` object is inside a `bar` directory. But that's all just presentation. Put another way: there's no such thing as an empty directory in S3.

So in this code, we're always just processing the one bucket, and waiting for manifest objects to appear in it.

> Then the only scenario where things could get out of order is if Scott's process created the buckets out of order - but that seems like something that is easily avoidable.

All that explanation above aside, I agree: as long as Scott always waits for one data set to be completely uploaded before starting the next, we're fine. This means he shouldn't ever start two data sets uploading in parallel, though.